src/ipc/ipc4/handler.c

```diff
@@ -975,6 +979,26 @@ void ipc_boot_complete_msg(struct ipc_cmd_hdr *header, uint32_t data)
 	header->ext = 0;
 }
 
+#ifdef CONFIG_LOG_BACKEND_ADSP_MTRACE
+void ipc_send_buffer_status_notify(void)
+{
+	/* TODO: handling notifies on non-zero cores not yet handled */
+	if (!cpu_is_me(0))
+		return;
+
+	/* TODO: need to find a cleaner solution */
+	if (msg_data.delayed_reply || ipc_cmd_in_progress)
+		return;
```

Comment:
Ok this is the ugliest part. If a notification is sent while kernel is waiting for a response, that response is not correctly handled. Not sure if this is a kernel bug, or something we have to avoid as per the protocol. @ujfalusi @RanderWang @mmaka1 @marcinszkudlinski ?

Reply:
Hrm, the kernel should be able to handle it and afaik I have seen it with closed source firmware. The tx+reply handling is detached from notification handling on the kernel side at least. It must be working.

Reply:
? is there any issue for bi-direction ipc msg ? I don't get it.   There are two sets of ipc register for tx and rx

Reply:
@ujfalusi @RanderWang Without this adhoc protection, I see BIND fail when a notification is send between BIND REQ received and before a response is sent back. I've not yet debugged why this happens and whether this is a problem on kernel or firmware side.

Reply:
@kv2019i, do you have logs from kernel (with sof_debug=513)?

Reply:
@RanderWang @ujfalusi Mystery solved. The problem was caused by sending a notify from a different DSP core, not related to the ongoing request-response. Timing just happened to match with that. Will send an updated PR today.